### PR TITLE
Adds ScheduledExecutorService for Polling the RSS feed

### LINKED ISSUE #2081: Add Polling to enable reading the feed URL periodically
Make use of Java's `ScheduledExecutorService` to read the RSS feed URL at the interval of `pollingFrequency`

### PR DESCRIPTION
**Pull request by sshivanii**
### Description
- This PR adds the Java `ScheduledExecutorService` to schedule reading the RSS Feed URL every 5 minutes. It is a draft to gather any feedback about using the Executor service.
- Currently, the unit test is reading from a real RSS feed URL because storing the URL in local files under `test/resources` directory resulted in `Invalid URI scheme file` errors. Finding the fix for this is still underway.
- DCO is failing, working on amending the commit messages.
 
### Issues Resolved
#2081 
 
### Check List
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by dlvenable — COMMENTED** [review]
I made a few comments on the main code.

File: data-prepper-plugins/rss-source/src/main/java/org/opensearch/dataprepper/plugins/source/rss/RSSSource.java [hunk 1]
```diff
@@ -32,41 +36,48 @@ public class RSSSource implements Source<Record<Document>> {
...
     }
 
     @Override
-    public void start(Buffer buffer) {
-        // TODO: Add to buffer
+    public void start(final Buffer<Record<Document>> buffer) {
         if (buffer == null) {
             throw new IllegalStateException("Buffer is null");
         }
+        final Runnable runnable = () -> {
```
[the remark this answers is not in the document]
Comment (sshivanii) [reply data-prepper-plugins/rss-source/src/main/java/org/opensearch/dataprepper/plugins/source/rss/RSSSource.java]: Nice suggestion, I'll take it!

File: data-prepper-plugins/rss-source/build.gradle [hunk 2]
```diff
@@ -18,6 +18,7 @@ dependencies {
     implementation 'com.fasterxml.jackson.core:jackson-core'
     implementation 'com.fasterxml.jackson.core:jackson-databind'
     implementation 'com.apptasticsoftware:rssreader:3.2.5'
+    implementation 'com.google.code.gson:gson:2.8.7' // TODO: Pretty print Items for demo, remove later
```
Comment (kkondaka) [inline data-prepper-plugins/rss-source/build.gradle:21]: This should be removed now, right?

File: data-prepper-plugins/rss-source/src/main/java/org/opensearch/dataprepper/plugins/source/rss/RssReaderTask.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    private final RssReader rssReader;
+
+    private final RSSSourceConfig rssSourceConfig;
+
+    private final Buffer<Record<Document>> buffer;
+
+    final Collection<Record<Document>> collection = new HashSet<>();
+
+
+    public RssReaderTask(RssReader rssReader, final RSSSourceConfig rssSourceConfig, final Buffer<Record<Document>> buffer) {
```
Comment (kkondaka) [inline data-prepper-plugins/rss-source/src/main/java/org/opensearch/dataprepper/plugins/source/rss/RssReaderTask.java:37]: It is probably good idea to just pass the URL here instead of rssSourceConfig as the argument?

File: data-prepper-plugins/rss-source/build.gradle (hunk 2 above)
Comment (sshivanii) [reply data-prepper-plugins/rss-source/build.gradle:21]: I'm keeping it for future intermediate demos.

File: data-prepper-plugins/rss-source/src/main/java/org/opensearch/dataprepper/plugins/source/rss/RssReaderTask.java (hunk 3 above)
Comment (sshivanii) [reply data-prepper-plugins/rss-source/src/main/java/org/opensearch/dataprepper/plugins/source/rss/RssReaderTask.java:37]: That would not be accurate. There's also `pollingFrequency` which by default will be 5 minutes. But it would be an optional argument the user can configure. I'll be adding it in a future PR.

Comment (kkondaka) [reply data-prepper-plugins/rss-source/src/main/java/org/opensearch/dataprepper/plugins/source/rss/RssReaderTask.java:37]: That may be true but I think, since this is an independent class, making it dependent on some thing like RSSSourceConfig is not ideal. Any necessary info should be explicitly passed instead of passing the entire config. I am not strict about this. Just a suggestion. What do you think @dlvenable

**Review by kkondaka — APPROVED** [review]
Apart from a minor comment, the rest looks good.

File: data-prepper-plugins/rss-source/build.gradle (hunk 2 above)
Comment (dlvenable) [reply data-prepper-plugins/rss-source/build.gradle:21]: I think we should remove it now. A couple reasons:

1. You can use Jackson for printing demo data
2. I don't even think it is being used
3. We might forget to remove it

File: data-prepper-plugins/rss-source/src/main/java/org/opensearch/dataprepper/plugins/source/rss/RssReaderTask.java (hunk 3 above)
Comment (dlvenable) [reply data-prepper-plugins/rss-source/src/main/java/org/opensearch/dataprepper/plugins/source/rss/RssReaderTask.java:37]: I tend to agree with @kkondaka  here. I think the RSSSource can just pass in what RssReaderTask needs. But, I also don't feel strong about this.

Also, `pollingFrequency` should never be needed here since the scheduling happens in RSSSource.

One configuration that might come up is the write timeout.

File: data-prepper-plugins/rss-source/src/test/java/org/opensearch/dataprepper/plugins/source/rss/RSSSourceTest.java [hunk 4]
```diff
@@ -32,54 +33,35 @@ class RSSSourceTest {
...
-    void test_when_extractItemsFromRssFeed_captured() {
-        ArgumentCaptor<RSSSourceConfig> argumentCaptor = ArgumentCaptor.forClass(RSSSourceConfig.class);
-        when(rssSourceConfig.getUrl()).thenReturn(VALID_RSS_URL);
-        doNothing().when(rssSource).extractItemsFromRssFeed(argumentCaptor.capture());
-        rssSource.extractItemsFromRssFeed(rssSourceConfig);
-        assertEquals(rssSourceConfig, argumentCaptor.getValue());
-    }
-
-    @ParameterizedTest
-    @ValueSource(strings = { "https://www.yahoo.com/news/rss/mostviewed", "https://forum.opensearch.org/latest.rss",
```
Comment (dlvenable) [inline data-prepper-plugins/rss-source/src/test/java/org/opensearch/dataprepper/plugins/source/rss/RSSSourceTest.java:60]: You might consider adding some of these old tests to a new test suite for integration tests: `RssSourceIT`. I think this could be a different PR.

File: data-prepper-plugins/rss-source/src/main/java/org/opensearch/dataprepper/plugins/source/rss/RssReaderTask.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+class RssReaderTask implements Runnable {
```
Comment (dlvenable) [inline data-prepper-plugins/rss-source/src/main/java/org/opensearch/dataprepper/plugins/source/rss/RssReaderTask.java:24]: This class should have a unit test. And it should be relatively easy to add it. You can mock the `RssReader`, `RSSSourceConfig`, and `Buffer`. 

Some test cases that I see off the bat:

* When items are empty, the buffer is not written to (this would fail as the code currently is)
* When one item is present, it is written to the buffer
* When multiple times are present, they are all written to the buffer
* Throws exception and does not write to the buffer if the reader throws an exception
* Failing to convert an item to the Event model does not write to the buffer, but the other items from that same read do write to the buffer

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/rss-source/build.gradle (hunk 2 above)
Comment (sshivanii) [reply data-prepper-plugins/rss-source/build.gradle:21]: It was being used in the `RSSSource.java` while creating the documents, I was printing them using GSON. But I'll look at other alternatives.

File: data-prepper-plugins/rss-source/src/test/java/org/opensearch/dataprepper/plugins/source/rss/RSSSourceTest.java (hunk 4 above)
Comment (sshivanii) [reply data-prepper-plugins/rss-source/src/test/java/org/opensearch/dataprepper/plugins/source/rss/RSSSourceTest.java:60]: Yes that's the plan. I removed these from unit test and intend to add them as integration tests in a future PR.

File: data-prepper-plugins/rss-source/src/main/java/org/opensearch/dataprepper/plugins/source/rss/RssReaderTask.java (hunk 3 above)
Comment (sshivanii) [reply data-prepper-plugins/rss-source/src/main/java/org/opensearch/dataprepper/plugins/source/rss/RssReaderTask.java:37]: Understood, I decided to update this.

File: data-prepper-plugins/rss-source/src/main/java/org/opensearch/dataprepper/plugins/source/rss/RssReaderTask.java (hunk 5 above)
Comment (sshivanii) [reply data-prepper-plugins/rss-source/src/main/java/org/opensearch/dataprepper/plugins/source/rss/RssReaderTask.java:24]: Considering that I decided to pass only the URL to the `RssReaderTask` constructor and not the entire `RSSSourceConfig`, mocking the config would not be needed anymore. Do we still want explicit testcases for this class?

Comment (kkondaka) [reply data-prepper-plugins/rss-source/src/main/java/org/opensearch/dataprepper/plugins/source/rss/RssReaderTask.java:24]: Yes, test cases are still needed, whether you mock the config or not. You would still have to mock RssReader and Buffer

**Review by kkondaka — APPROVED** [review]

**Review by dlvenable — APPROVED** [review]
